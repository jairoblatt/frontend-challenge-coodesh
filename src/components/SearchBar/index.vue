<template>
  <div class="search-bar">
    <input class="search-bar__input" type="text" placeholder="Search user:" @input="onInput" />
    <div class="search-bar__prepend">
      <Icon size="28">mdi-account</Icon>
    </div>
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";

export default Vue.extend({
  props: {
    debounceDelay: {
      type: Number,
      default: 400,
    } as PropOptions<number>,
  },

  data: () => ({
    timeoutId: 0,
  }),

  methods: {
    onInput($event: Event) {
      clearTimeout(this.timeoutId);
      this.timeoutId = setTimeout(() => {
        this.$emit("input", ($event.target as HTMLInputElement)?.value);
      }, this.debounceDelay);
    },
  },
});
</script>

<style lang="postcss" scoped>
.search-bar {
  @apply w-full border border-gray-100 bg-white rounded-md pl-5 flex justify-between items-center overflow-hidden transition-colors;
}

.search-bar__input {
  @apply outline-none h-full pl-5 w-full text-light-onSurfacePrimary font-semibold;
}

.search-bar__prepend {
  @apply border border-gray-100 flex justify-center;
}

.search-bar__prepend >>> i {
  @apply text-light-onSurfacePrimary;
}

.search-bar:focus-within {
  @apply border-primary-light;
}

.search-bar:focus-within .search-bar__prepend {
  @apply bg-primary-light border-primary-light;
}

.search-bar:focus-within .search-bar__prepend >>> i {
  @apply text-white;
}
</style>
