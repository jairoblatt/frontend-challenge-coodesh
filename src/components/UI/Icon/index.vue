<template>
  <i v-if="iconSanitize" :class="`mdi ${iconSanitize}`" :style="styles"></i>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Icon",

  props: {
    size: {
      type: [String, Number],
      default: 14,
    } as PropOptions<string | number>,

    color: String as PropType<string>,

    icon: {
      type: String,
      default: "",
    } as PropOptions<string>,
  },

  computed: {
    iconSanitize(): string {
      return this.$slots.default?.[0]?.text || this.icon;
    },

    styles(): CSS.Properties {
      return {
        fontSize: this.size + "px",
        ...(this.color && { color: this.color }),
      };
    },
  },
});
</script>
