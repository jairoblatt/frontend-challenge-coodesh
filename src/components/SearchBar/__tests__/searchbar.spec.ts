import { mount, Wrapper, MountOptions, createLocalVue } from "@vue/test-utils";
import SearchBar from "@/components/SearchBar/index.vue";
import Icon from "@/components/UI/Icon/index.vue";

describe("SearchBar.vue ", () => {
  const localVue = createLocalVue();
  localVue.component("Icon", Icon);

  type Instance = InstanceType<typeof SearchBar>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(SearchBar, {
        ...options,
        localVue,
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should emit "input" event when input element is typed', () => {
    jest.useFakeTimers();

    const wrapper = mountFunction({
      propsData: {
        debounceDelay: 0,
      },
    });

    wrapper.find("input").setValue("Something");

    jest.advanceTimersByTime(100);

    expect(wrapper.emitted("input")).toBeTruthy();
    expect(wrapper.emitted("input")?.length).toBe(1);
    expect(wrapper.emitted("input")?.[0]).toEqual(["Something"]);
  });

  it('Should render "Icon" component with the mdi-account icon', () => {
    const wrapper = mountFunction();

    const iconComponent = wrapper.findComponent(Icon);

    expect(iconComponent).toBeDefined();
  });
});
