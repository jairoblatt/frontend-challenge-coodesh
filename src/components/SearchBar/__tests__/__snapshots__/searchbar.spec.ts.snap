// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar.vue  Should mount the component 1`] = `
"<div class=\\"search-bar\\"><input type=\\"text\\" placeholder=\\"Search user:\\" class=\\"search-bar__input\\">
  <div class=\\"search-bar__prepend\\"><i class=\\"mdi mdi-account\\" style=\\"font-size: 28px;\\"></i></div>
</div>"
`;
